client/src/game/reducers/noteIdentity.ts

```diff
@@ -47,91 +47,88 @@ export function parseIdentity(variant: Variant, keyword: string): CardIdentity {
+  const positiveIdent = [];
+  const negativeIdent = [];
+  let positiveRanks = new Set<number>();
+  let positiveSuits = new Set<number>();
+  for (const substring of keywordPreTrim.split(",")) {
+    const trimmed = substring.trim();
+    const negative = trimmed.startsWith("!");
+    const identity = parseIdentity(
+      variant,
+      (negative ? trimmed.substring(1) : trimmed).trim(),
+    );
+    if (negative) {
+      negativeIdent.push(identity);
+    } else if (identity.rank === null) {
+      if (identity.suitIndex === null) {
+        return null;
       }
-    } else if (identity.suitIndex !== null && identity.rank === null) {
-      // Encountered a suit item, expand to all cards of that suit
-      for (const rank of variant.ranks) {
-        // Check that this identity is not already present in the list
-        if (
-          !possibilities.some(
-            (possibility) =>
-              possibility[0] === identity.suitIndex && possibility[1] === rank,
-          )
-        ) {
-          possibilities.push([identity.suitIndex, rank]);
+      positiveSuits.add(identity.suitIndex);
+    } else if (identity.suitIndex === null) {
+      positiveRanks.add(identity.rank);
+    } else {
+      positiveIdent.push(identity);
+    }
+  }
+  // Start with the cross of the positive ranks and suits.
+  const hasSuits = positiveSuits.size > 0;
+  if (!positiveSuits.size) positiveSuits = new Set([0, 1, 2, 3, 4, 5]);
+  if (hasSuits && !positiveRanks.size) positiveRanks = new Set([1, 2, 3, 4, 5]);
+  const zeros = [0, 0, 0, 0, 0, 0];
+  const positiveSuitsTemplate = zeros.slice();
+  positiveSuits.forEach((suit) => {
+    positiveSuitsTemplate[suit] = 1;
+  });
+  const identityMap = [];
+  for (let rank = 1; rank <= 5; ++rank) {
+    identityMap.push(
+      positiveRanks.has(rank) ? positiveSuitsTemplate.slice() : zeros.slice(),
+    );
+  }
+  // Then add individual items and remove all negatives.
+  for (const identities of [positiveIdent, negativeIdent]) {
+    const negative = identities === negativeIdent;
+    for (const identity of identities) {
+      for (let rank = 1; rank <= 5; ++rank) {
+        if (identity.rank !== null && identity.rank !== rank) {
+          continue;
         }
-      }
-    } else if (identity.suitIndex === null && identity.rank !== null) {
-      // Encountered a rank item, expand to all cards of that rank
-      for (let suitIndex = 0; suitIndex < variant.suits.length; suitIndex++) {
-        // Check that this identity is not already present in the list
-        if (
-          !possibilities.some(
-            (possibility) =>
-              possibility[0] === suitIndex && possibility[1] === identity.rank,
-          )
-        ) {
-          possibilities.push([suitIndex, identity.rank]);
+        for (let suitIndex = 0; suitIndex < variant.suits.length; ++suitIndex) {
+          if (identity.suitIndex !== null && identity.suitIndex !== suitIndex) {
+            continue;
```

Comment:
instead of using a continue statement, invert the conditional and stick the line inside

Reply:
I like `continue` because it means less nesting, so IMHO lower (visual) complexity.

Reply:
i dont understand how it results in less nesting.
the cyclomatic complexity of the function is the same.